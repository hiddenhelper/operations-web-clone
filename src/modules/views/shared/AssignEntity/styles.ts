import { createStyles, makeStyles } from '@material-ui/core/styles';
import { STYLE } from '../../../../constants';
import { toREM } from '../../../../utils/generalUtils';

export const useStyles = makeStyles(theme =>
  createStyles({
    autoCompleteWrapper: {
      '& .MuiSelect-root': {
        width: '85%',
      },

      '& .MuiAutocomplete-popper': {
        '& .MuiAutocomplete-paper': {
          boxShadow: '0 0 5px 0 rgba(0,0,0,0.15)',
          border: `1px solid ${STYLE.COLOR.LIGHT_GRAY3}`,
          borderRadius: `${toREM(5)}`,
        },

        '& .MuiPaper-root': {
          padding: 0,
        },

        '& .MuiAutocomplete-listbox': {
          padding: 0,

          '& .MuiAutocomplete-option': {
            height: `${toREM(42)}`,
            fontSize: `${toREM(STYLE.FONT.SIZE.PARAGRAPH_BIG)}`,
            borderBottom: `1px solid ${STYLE.COLOR.QUINARY_LIGHTER}`,
          },

          '& li:last-child': {
            borderBottom: 'none',
          },

          '& [data-testid="search-option-item"]': {
            width: '100%',
            color: STYLE.COLOR.ACCENT_QUINARY,
            maxWidth: '95%',
            whiteSpace: 'nowrap',
            overflow: 'hidden',
            textOverflow: 'ellipsis',
          },

          '& .MuiAutocomplete-option[data-focus="true"]': {
            '&:hover': {
              backgroundColor: 'transparent !important',
            },
          },
        },
      },
      '& .MuiAutocomplete-root': {
        backgroundColor: STYLE.COLOR.INPUT_BACKGROUND,

        '& .MuiAutocomplete-inputRoot[class*="MuiOutlinedInput-root"]': {
          padding: 0,
          paddingLeft: '9px',
          backgroundColor: STYLE.COLOR.INPUT_BACKGROUND,
        },
      },
      '& .MuiButton-root:hover': {
        backgroundColor: 'transparent',
      },
    },
    autoCompleteButton: {
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      width: '100%',
      position: 'relative',
    },
    createButtonWrapper: {
      position: 'relative',
    },
    createButton: {
      fontSize: '13px',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      width: '100%',
      position: 'relative',
      color: STYLE.COLOR.ACCENT_PRIMARY,
      fontWeight: 600,
      textTransform: 'inherit',
      left: '0px',
      paddingLeft: '0 !important',
    },
    errorInput: {
      '& .MuiOutlinedInput-notchedOutline': {
        borderColor: STYLE.COLOR.ERROR_FORM,
      },
    },
  })
);
