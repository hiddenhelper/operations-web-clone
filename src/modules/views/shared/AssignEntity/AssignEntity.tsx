import React, { memo, useEffect, useMemo, useCallback, useState } from 'react';

import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Popper from '@material-ui/core/Popper';
import Autocomplete from '@material-ui/lab/Autocomplete';

import ControlledInput from '../FormHandler/ControlledInput';

import { GeneralModel } from '../../../models';
import { useDebounce } from '../../../../utils/useDebounce';
import { getConditionalDefaultValue } from '../../../../utils/generalUtils';
import { useStyles } from './styles';

export interface IAssignEntityProps {
  index: number;
  tempId: string;
  result: any[];
  params?: GeneralModel.IQueryParams;
  optionLabel: string;
  assignValue: any;
  isLoading: boolean;
  existRelation: boolean;
  showCreateNew?: boolean;
  placeholder?: string;
  inputLabel?: string;
  showError?: boolean;
  disableClearable?: boolean;
  styleClass?: string;
  disabled?: boolean;
  isRequired?: boolean;
  renderOption: (option: any, inputValue: string) => React.ReactNode;
  search: (params: GeneralModel.IQueryParams, tempId: string) => void;
  onCreate?: (searchText: string, tempId: string) => void;
  onSelect: (index: number, item: any, tempId: string) => void;
  onReset: (index: number, tempId: string) => void;
}

const AssignEntity = ({
  index,
  tempId,
  assignValue,
  result,
  params = {},
  disabled = false,
  optionLabel,
  isLoading,
  placeholder = '',
  showCreateNew = false,
  existRelation,
  showError,
  disableClearable = true,
  inputLabel = 'Client Name',
  styleClass = '',
  isRequired,
  search,
  onSelect,
  onCreate,
  renderOption,
  onReset,
}: IAssignEntityProps) => {
  const classes = useStyles();
  const [searchText, setSearch] = useState<string>('');
  const [currentRelation, setRelation] = useState<boolean>(existRelation);
  const debouncedSearch = useDebounce(searchText, 350);

  const query = useMemo(() => ({ isDeveloper: false, maxItems: 30, ...params }), [params]);
  const showPopper = useMemo(() => debouncedSearch.length >= 3, [debouncedSearch]);
  const resultList = useMemo(() => (showCreateNew && debouncedSearch.length >= 3 ? [...result, { [optionLabel]: '', isCreateNew: true }] : result), [
    showCreateNew,
    result,
    optionLabel,
    debouncedSearch,
  ]);

  const getOptionLabel = useCallback(option => (option && option[optionLabel] ? option[optionLabel] : ''), [optionLabel]);

  const onSearchChange = useCallback(
    event => {
      event.persist();
      setSearch(event.target.value);
    },
    [setSearch]
  );

  const filterOptionListWithCreate = useCallback(
    (optionList, value) => {
      return optionList.filter(opt => opt[optionLabel].toLowerCase().indexOf(value.inputValue.toLowerCase()) !== -1 || opt.isCreateNew);
    },
    [optionLabel]
  );

  const filterOptionList = useCallback(
    (optionList, value) => {
      return optionList.filter(opt => opt[optionLabel].toLowerCase().indexOf(value.inputValue.toLowerCase()) !== -1);
    },
    [optionLabel]
  );

  const filterOptions = useMemo(() => (showCreateNew ? filterOptionListWithCreate : filterOptionList), [
    showCreateNew,
    filterOptionListWithCreate,
    filterOptionList,
  ]);

  const onCreateHandler = useCallback(() => {
    onCreate(debouncedSearch, tempId);
  }, [onCreate, debouncedSearch, tempId]);

  const autocompleteInput = useCallback(
    inputParams => (
      <TextField
        onChange={onSearchChange}
        variant="outlined"
        autoComplete="off"
        placeholder={placeholder}
        fullWidth={true}
        inputProps={{
          'data-testid': 'assign-entity-input',
        }}
        {...inputParams}
      />
    ),
    [onSearchChange, placeholder]
  );

  const onCloseHandler = useCallback(
    (event, reason) => {
      if (reason === 'blur' && assignValue && assignValue.id === null) {
        onReset(index, tempId);
      }
    },
    [onReset, assignValue, index, tempId]
  );

  const onSelectItem = useCallback(
    (item, value) => {
      onSelect(index, value, tempId);
    },
    [tempId, index, onSelect]
  );

  const onRenderOption = useCallback(
    (opt, { inputValue }) => {
      if (opt.isCreateNew && showPopper) {
        return (
          <div className={classes.createButtonWrapper}>
            <Button data-testid="create-entity" className={classes.createButton} disableTouchRipple={true} onClick={onCreateHandler}>
              + Create New Client {debouncedSearch}
            </Button>
          </div>
        );
      }
      return renderOption(opt, inputValue);
    },
    [debouncedSearch, renderOption, onCreateHandler, showPopper, classes]
  );

  useEffect(() => {
    setRelation(existRelation);
  }, [existRelation]);

  useEffect(() => {
    if (currentRelation && !existRelation) setSearch('');
  }, [currentRelation, existRelation, setSearch]);

  useEffect(() => {
    if (debouncedSearch.length >= 3) search({ ...query, nameContains: debouncedSearch }, tempId);
  }, [search, debouncedSearch, tempId]); // eslint-disable-line

  return (
    <ControlledInput styleClass={`${classes.autoCompleteWrapper} ${styleClass}`} label={inputLabel} required={isRequired} showMark={isRequired}>
      <Autocomplete
        id="autocomplete-assign"
        data-testid="autocomplete-wrapper"
        disableClearable={disableClearable}
        options={resultList}
        value={assignValue}
        loading={isLoading}
        disabled={disabled}
        className={getConditionalDefaultValue(showError, classes.errorInput, '')}
        PopperComponent={props =>
          showPopper && (
            <Popper
              disablePortal={true}
              placement={'bottom-start'}
              {...props}
              style={{
                display: 'block',
                position: 'absolute',
                width: '100%',
                left: '0px',
              }}
            />
          )
        }
        onClose={onCloseHandler}
        freeSolo={true}
        onChange={onSelectItem}
        filterOptions={filterOptions}
        getOptionLabel={getOptionLabel}
        renderInput={autocompleteInput}
        renderOption={onRenderOption}
      />
    </ControlledInput>
  );
};

export default memo(AssignEntity);
