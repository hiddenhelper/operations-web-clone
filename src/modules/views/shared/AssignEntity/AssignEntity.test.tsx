import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import AssignEntity, { IAssignEntityProps } from './AssignEntity';

jest.useFakeTimers();

describe('AssignEntity Component', () => {
  let props: IAssignEntityProps;

  beforeEach(() => {
    props = {
      index: 0,
      tempId: 'temp-id-0',
      result: [{ name: 'test' }, { name: 'test 2' }],
      params: { isDeveloper: false },
      isLoading: false,
      existRelation: false,
      optionLabel: 'name',
      assignValue: '',
      renderOption: jest.fn(),
      search: jest.fn(),
      onCreate: jest.fn(),
      onSelect: jest.fn(),
      onReset: jest.fn(),
    };
  });

  it('should render', () => {
    const { container } = render(<AssignEntity {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should search', () => {
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test value search' } });
    });

    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(props.search).toHaveBeenCalledWith({ isDeveloper: false, nameContains: 'test value search', maxItems: 30 }, 'temp-id-0');
  });

  it('should render option list', () => {
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.mouseDown(input);
    });

    expect(props.renderOption).toHaveBeenCalled();
  });

  it('should render create new option', () => {
    props.showCreateNew = true;
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test value search' } });
    });

    act(() => {
      fireEvent.mouseDown(input);
    });

    act(() => {
      jest.runOnlyPendingTimers();
    });

    expect(wrapper.getByTestId('create-entity')).toBeTruthy();
  });

  it('should create new', () => {
    props.showCreateNew = true;
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test value search' } });
    });

    act(() => {
      fireEvent.mouseDown(input);
    });

    act(() => {
      jest.runOnlyPendingTimers();
    });

    const createButton = wrapper.getByTestId('create-entity');

    act(() => {
      fireEvent.click(createButton);
    });

    expect(props.onCreate).toHaveBeenCalledWith('test value search', 'temp-id-0');
  });

  it('should select', () => {
    props.showCreateNew = true;
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test' } });
    });

    act(() => {
      fireEvent.mouseDown(input);
    });

    act(() => {
      jest.runOnlyPendingTimers();
    });

    act(() => {
      fireEvent.click(wrapper.getAllByRole('option')[0]);
    });

    expect(props.onSelect).toHaveBeenCalledWith(0, { name: 'test' }, 'temp-id-0');
  });

  it('should reset', () => {
    props.assignValue = { id: null, name: 'test' };
    const wrapper = render(<AssignEntity {...props} />);

    const input = wrapper.getByTestId('autocomplete-wrapper').querySelector('input');

    act(() => {
      fireEvent.change(input, { target: { value: 'test' } });
    });

    act(() => {
      fireEvent.mouseDown(input);
    });

    act(() => {
      fireEvent.blur(input);
    });

    expect(props.onReset).toHaveBeenCalled();
  });

  it('should clear text when relation', () => {
    props.existRelation = true;
    const wrapper = render(<AssignEntity {...props} />);

    props.existRelation = false;
    wrapper.rerender(<AssignEntity {...props} />);

    expect(wrapper.container).toMatchSnapshot();
  });
});
