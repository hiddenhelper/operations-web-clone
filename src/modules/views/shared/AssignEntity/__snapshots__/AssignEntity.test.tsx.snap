// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssignEntity Component should clear text when relation 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-107 makeStyles-autoCompleteWrapper-99 "
  >
    <h1
      class="MuiTypography-root MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
    >
      Client Name
       
    </h1>
    <div>
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root"
        data-testid="autocomplete-wrapper"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              id="autocomplete-assign"
              placeholder=""
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-109 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-110"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AssignEntity Component should render 1`] = `
<div>
  <div
    class="MuiFormControl-root makeStyles-formControl-9 makeStyles-autoCompleteWrapper-1 "
  >
    <h1
      class="MuiTypography-root MuiTypography-caption MuiTypography-colorSecondary MuiTypography-alignLeft"
    >
      Client Name
       
    </h1>
    <div>
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root"
        data-testid="autocomplete-wrapper"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              id="autocomplete-assign"
              placeholder=""
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-12"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
